xonsh/parsers/base.py

```diff
@@ -2150,10 +2159,26 @@ def p_subproc_atom_uncaptured(self, p):
         p0._cliarg_action = 'splitlines'
         p[0] = p0
 
-    def p_subproc_atom_captured(self, p):
+    def p_subproc_atom_captured_stdout(self, p):
         """subproc_atom : dollar_lparen_tok subproc RPAREN"""
         p1 = p[1]
-        p0 = xonsh_call('__xonsh_subproc_captured__', args=p[2],
+        p0 = xonsh_call('__xonsh_subproc_captured_stdout__', args=p[2],
+                        lineno=p1.lineno, col=p1.lexpos)
+        p0._cliarg_action = 'splitlines'
+        p[0] = p0
+
+    def p_subproc_atom_captured_object(self, p):
+        """subproc_atom : bang_lparen_tok subproc RPAREN"""
+        p1 = p[1]
+        p0 = xonsh_call('__xonsh_subproc_captured_object__', args=p[2],
+                        lineno=p1.lineno, col=p1.lexpos)
+        p0._cliarg_action = 'splitlines'
+        p[0] = p0
+    
+    def p_subproc_atom_captured_hiddenobject(self, p):
+        """subproc_atom : bang_lbracket_tok subproc RBRACKET"""
+        p1 = p[1]
+        p0 = xonsh_call('__xonsh_subproc_captured_hiddenobject__', args=p[2],
```

Comment:
These two cases don't work, and I am not wholly sure what their purpose is except as potential setup for and/or behaviour.

```console
scopatz@artemis ~ $ echo !(echo hello)
Traceback (most recent call last):
  File "/home/scopatz/.local/lib/python3.5/site-packages/xonsh/base_shell.py", line 150, in default
    self.execer.exec(code, mode='single', glbs=self.ctx)  # no locals
  File "/home/scopatz/.local/lib/python3.5/site-packages/xonsh/execer.py", line 126, in exec
    return exec(code, glbs, locs)
  File "<xonsh-code>", line 1, in <module>
AttributeError: 'CompletedCommand' object has no attribute 'splitlines'
scopatz@artemis ~ $ echo ![echo hello]
hello
Traceback (most recent call last):
  File "/home/scopatz/.local/lib/python3.5/site-packages/xonsh/base_shell.py", line 150, in default
    self.execer.exec(code, mode='single', glbs=self.ctx)  # no locals
  File "/home/scopatz/.local/lib/python3.5/site-packages/xonsh/execer.py", line 126, in exec
    return exec(code, glbs, locs)
  File "<xonsh-code>", line 1, in <module>
AttributeError: 'HiddenCompletedCommand' object has no attribute 'splitlines'
```